packages/runtime/container-runtime/src/summary/documentSchema.ts

```diff
@@ -551,7 +559,9 @@ export class DocumentsSchemaController {
 	 * @returns Optional message to send.
 	 */
 	public maybeSendSchemaMessage(): IDocumentSchemaChangeMessage | undefined {
-		if (this.sendOp && this.futureSchema !== undefined) {
+		const sendOp = this.sendOp;
+		this.sendOp = false;
```

Comment:
Shouldn't this be done inside the if condition below? Otherwise, if `this.futureSchema` is undefined, the schema op won't be sent and `this.sendOp` will be false. is that expected? Can't `this.futureSchema` change later?

Reply:
Not today. Current design - client only tries to change schema once in a session. Basically, from statistical POV, I do not care about optimizing a case where one client wants compression, but another wants op grouping, and they happen ops at the same time - one of them will win the race. The other client will get his chance in next session.
It just not worth it extra complexity (and possibly bugs) - system will always get eventually to final state, it might take a bit longer, that's it.

See this note at the top of constructor (I'll duplicate it above this function as well):
> \* When a number of clients race to change a schema, then only one of them will win, all others will fail because they will
> \* reference old schema that is no longer in effect.
> \* Clients can retry, but current implementation is simply - they will not (and will rely on next session / reload to do
> \* recalc and decide if schema needs to be changed or not).

Reply:
But I'll move reset in, just in case, to be prepared for any future changes :)

Reply:
A conatiner can try to change schema only once but can it respond to a schema change more than once in a session?
For example, say id compressor is supported but disabled in client1 running 2.0. Client2 loads with 3.0 and sends a schema change op with id compressor enabled, will client1 (2.0) enable id compressor in the same session or will it shut down and when it restarts, it will catch up, see the schema change op and enable id compressor on load?

Reply:
Yes, it processes all the ops and reacts to them.
1. There is some coverage in "Existing document, changes required; race conditions" test
2. I'm also planning to add end-to-end mini stress test to be absolutely sure :) I have built it as part of this investigation, but it felt a bit too heavy, need to clean it up a bit to make more inclusive (those grouped ops!) but also not as slow (right now it takes forever to run).
3. Our service tests also exercise that flow continuously (though some of that coverage is indirect, i.e. we know that system is eventually consistent only because clients behave as expected, for example - they assert if they see ID compressor ops but schema says - it's off for them).